src/bundles/experiments.js

```diff
@@ -0,0 +1,122 @@
+import { createSelector } from 'redux-bundler'
+
+export const ACTIONS = {
+  EXP_TOGGLE_STARTED: 'EXPERIMENTS_TOGGLE_STARTED',
+  EXP_TOGGLE_FINISH: 'EXPERIMENTS_TOGGLE_FINISH',
+  EXP_TOGGLE_FAILED: 'EXPERIMENTS_TOGGLE_FAILED',
+  EXP_UPDATE_STATE: 'EXPERIMENTS_UPDATE_STATE'
+}
+
+const EXPERIMENTS = [
+  {
+    key: 'npmOnIpfs',
+    actionUrls: [
+      {
+        url: 'https://github.com/ipfs-shipyard/npm-on-ipfs',
+        key: 'readMoreUrl'
+      },
+      {
+        url: 'https://github.com/ipfs-shipyard/npm-on-ipfs/issues',
+        key: 'issueUrl'
+      },
+      {
+        url: 'https://github.com/ipfs-shipyard/npm-on-ipfs',
+        key: 'feedbackUrl'
+      }
+    ],
+    desktop: true
+  }
+]
+
+const mergeState = (state, payload) => Object.keys(payload).reduce(
+  (all, key) => ({
+    ...all,
+    [key]: {
+      ...state[key],
+      ...payload[key]
+    }
+  }),
+  state
+)
+
+const toggleEnabled = (state, key) => {
+  return unblock({
+    ...state,
+    [key]: {
+      ...state[key],
+      enabled: !(state && state[key] && state[key].enabled)
+    }
+  }, key)
+}
+
+const unblock = (state, key) => {
+  return {
+    ...state,
+    [key]: {
+      ...state[key],
+      blocked: false
+    }
+  }
+}
+
+const block = (state, key) => {
+  return {
+    ...state,
+    [key]: {
+      ...state[key],
+      blocked: true
+    }
+  }
+}
+
+export default {
+  name: 'experiments',
+
+  persistActions: [
+    ACTIONS.EXP_TOGGLE_FINISH,
+    ACTIONS.EXP_TOGGLE_FAILED,
+    ACTIONS.EXP_UPDATE_STATE
+  ],
+
+  reducer: (state = {}, action) => {
+    if (action.type === ACTIONS.EXP_TOGGLE_STARTED) {
+      return block(state, action.payload.key)
+    }
+
+    if (action.type === ACTIONS.EXP_UPDATE_STATE) {
+      return mergeState(state, action.payload)
+    }
+
+    if (action.type === ACTIONS.EXP_TOGGLE_FINISH) {
+      return toggleEnabled(state, action.payload.key)
+    }
+
+    if (action.type === ACTIONS.EXP_TOGGLE_FAILED) {
+      return unblock(state, action.payload.key)
+    }
+
+    return state
+  },
+
+  doExpToggleAction: key => async ({ dispatch }) => {
+    if (!key) return
+
+    dispatch({
+      type: ACTIONS.EXP_TOGGLE_STARTED,
+      payload: { key }
+    })
+  },
+
+  selectExperimentsState: state => state.experiments,
+
+  selectExperiments: createSelector(
+    'selectIsIpfsDesktop',
+    'selectExperimentsState',
+    (isDesktop, state) => EXPERIMENTS
+      .filter(e => !!e.desktop === isDesktop)
```

Comment:
I like the neat code, but I think we need more flexibility here. My assumption here is that all experiments will be available on desktop and some will be available on desktop and webui. We are deliberately hiding all experiments from webui until we get an api that lets us trigger a restart of the daemon, but once we have that, it's be nice to let webui toggle things like pubsub, mdns etc.

Reply:
We're just hiding desktop-specific experiments when we're not inside IPFS Desktop. What's the issue? Didn't understand

Reply:
Yep, this seems to be functioning correctly. It will display everything in desktop and filter out desktop explicit experiments from any other context.

Reply:
You're right. I was thinking about a situtation (but not saying!) where we want to show an experiment on webui but not desktop... and that's probably not going to happen, so this is fine.

Could we change the key on the experiments label from `desktop: true` to `desktopOnly: true`?

Reply:
Good call, I'll update that 👌